parsons/catalist/catalist.py

```diff
@@ -0,0 +1,388 @@
+        if match:
+            actions.append("match")
+        if export:
+            actions.append("export")
+        action = urllib.parse.quote(",".join(actions))
+
+        endpoint_params = ["action", action]
+
+        if isinstance(file_ids, list):
+            encoded_files = urllib.parse.quote(",".join(file_ids))
+        else:
+            encoded_files = file_ids
+
+        endpoint_params.extend(["file", encoded_files])
+
+        endpoint = "/".join(endpoint_params)
+
+        logger.debug(f"Executing request to endpoint {self.connection.uri + endpoint}")
+
+        query_params = {"token": self.token}
+        if copy_to_sandbox:
+            query_params["copyToSandbox"] = "true"
+        if export_filename_suffix:
+            query_params["subClientName"] = export_filename_suffix
+
+        endpoint = endpoint + "?" + urllib.parse.urlencode(query_params)
+
+        result = self.connection.get_request(endpoint)
+
+        return result
+
+    def status(self, id: str) -> dict:
+        """Check status of a match job."""
+        endpoint = "/".join(["status", "id", id])
+        query_params = {"token": self.token}
+        result = self.connection.get_request(endpoint, params=query_params)
+        return result
+
+    def await_completion(self, id: str, wait: int = 30) -> Table:
+        """Await completion of a match job. Return matches when ready.
+
+        This method will poll the status of a match job on a timer until the job is
+        complete. By default, polls once every 30 seconds.
+
+        Note that match job completion can take from 10 minutes up to 6 hours or more
+        depending on concurrent traffic. Consider your strategy for polling for
+        completion."""
+        while True:
+            response = self.status(id)
+            status = response["process"]["processState"]
+            if status in ("Finished", "Error", "Stopped", "Exception"):
+                logger.info(f"Job {id} is complete with status {status}.")
+                break
+
+            logger.info(f"Job {id} has status {status}, awaiting completion.")
+            time.sleep(wait)
+
+        if status != "Finished":
+            raise RuntimeError(f"Failed to successfully run match job. [job={id}]")
+
+        result = self.load_matches(id)
+        return result
+
+    def load_matches(self, id: str) -> Table:
+        """Take a completed job ID, download and open the match file as a Table.
+
+        Result will be a Table with all the original columns along with columns 'DWID',
+        'CONFIDENCE', 'ZIP9', and 'STATE'. The original column headers will be prepended
+        with 'COL#-'."""
+        # Validate that the job is complete
+        response = self.status(str(id))
+
+        try:
+            assert response["process"]["processState"] == "Finished"
+            logger.info(f"Validated that job {id} completed successfully.")
+        except AssertionError:
+            raise AssertionError(
```

Comment:
Is there any extra level of detail we can provide here as to why the job failed? Even including the error code from the response might be useful

Reply:
Asking Catalist for this

Reply:
Done with a37527a99c2ace60e26e30dc9cc59bfc141a3967